### PR TITLE
Created request and response objects for sdk 2.x

### PR DESCRIPTION
**Pull request by lalwani**
Added

- AuthContext
- AuthDestination
- AuthResponse
- AuthType
- CrossApp
- PrefillInfo
- AuthException

### CODE REVIEW

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/exception/AuthException.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.uber.sdk2.auth.api.exception
+
+/** Represents the exception that occurred during the authentication request. */
+sealed class AuthException(override val message: String) : Throwable(message) {
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/api/exception/AuthException.kt:19]: I think we should not extend from `Throwable`. It is the top type and encompasses all `Error` which are always unrecoverable (e.g. `StackOverflowError`.

Should we extend from `RuntimeException`?

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/AuthDestination.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ */
+package com.uber.sdk2.auth.api.request
+
+/** Represents the destination app to authenticate the user. */
+sealed class AuthDestination {
+  /**
+   * Authenticating within the same app by using a system webview, a.k.a Custom Tabs. If custom tabs
+   * are unavailable the authentication flow will be launched in the system browser app.
+   */
+  object InApp : AuthDestination()
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/AuthDestination.kt:24]: `data object`

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/AuthType.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * Represents the type of authentication to perform.
+ *
+ * @see AuthContext
+ */
+sealed class AuthType {
+  /** The authorization code flow. */
+  object AuthCode : AuthType()
+
+  /** The proof key for code exchange (PKCE) flow. This is the recommended flow for mobile apps. */
+  object PKCE : AuthType()
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/AuthType.kt:28]: `data object`

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/CrossApp.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+/** Provides different apps that could be used for authentication using SSO flow. */
+sealed class CrossApp {
+  /** The Eats app. */
+  object Eats : CrossApp()
+
+  /** The Rider app. */
+  object Rider : CrossApp()
+
+  /** The Driver app. */
+  object Driver : CrossApp()
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/CrossApp.kt:27]: `data object`

**Review by psteiger — COMMENTED** [review]
I have a personal preference of grouping the top-level definitions in a single file if they're small and closely related to each other. But that's just preference.

**Comment by lalwani** [comment]
> I have a personal preference of grouping the top-level definitions in a single file if they're small and closely related to each other. But that's just preference.

I'd like to keep it this way. To me it becomes easier in searching, referencing and more readable. Thanks for the suggestion.

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/exception/AuthException.kt (hunk 1 above)
Comment (lalwani) [reply authentication/src/main/kotlin/com/uber/sdk2/auth/api/exception/AuthException.kt]: Good call

**Review by psteiger — APPROVED** [review]

**Pull request merged**